packages/protobuf/src/reflect/nested-types.ts

```diff
@@ -43,3 +48,34 @@ export function* nestedTypes(
       break;
   }
 }
+
+/**
+ * Returns the ancestors of a given Protobuf element, up to the file.
+ */
+export function parentTypes(desc: AnyDesc): Parent[] {
+  const parents: Parent[] = [];
+  while (desc.kind !== "file") {
+    const p = parent(desc);
+    desc = p;
+    parents.push(p);
+  }
+  return parents;
+}
+
+type Parent = DescFile | DescEnum | DescMessage | DescService;
+
+function parent(desc: Exclude<AnyDesc, DescFile>): Parent {
+  switch (desc.kind) {
+    case "enum_value":
+    case "field":
+    case "oneof":
+    case "rpc":
+      return desc.parent;
+    case "service":
+      return desc.file;
+    case "extension":
+    case "enum":
+    case "message":
+      return desc.parent ?? desc.file;
```

Comment:
I wonder whether we should look into using `parent: DescMessage | DescFile` instead.

Reply:
I took a note to consider this.